src/azul/chalice.py

```diff
@@ -145,3 +159,122 @@ def _log_response(self, response):
+        spec_validators: Dict[str, SpecValidation] = dict()
+        for parameter in self._extract_parameters(resource_path, http_method):
+            temp = SpecValidation(**parameter)
+            if 'schema' in parameter:
+                parameter_schema = parameter['schema']
+            elif 'content' in parameter:
+                for k, v in parameter['content'].items():
+                    parameter_schema = v['schema']
+            temp.validator = schema.type_lookup(parameter_schema['type'])
+            spec_validators[temp.name] = temp
+        return spec_validators
+
+    def validate_params(self) -> None:
+        def fmt_error(err_description, parameters):
+            joined = ', '.join(f'{p!r}' for p in parameters)
+            return f'{err_description} {pluralize("parameter", len(parameters))} {joined}'
+
+        # OpenAPI requires HTTP method names be lower case
+        http_method = self.current_request.method.lower()
+        resource_path = self.current_request.context['resourcePath']
+
+        query_params = self.current_request.query_params if self.current_request.query_params else {}
+        uri_params = self.current_request.uri_params if self.current_request.uri_params else {}
+        all_provided_params = {**query_params, **uri_params}
+
+        provided_params = set(all_provided_params.keys())
+        mandatory_params = set()
+        validation_params = set()
+
+        spec_parameters = self.create_spec_validators(resource_path, http_method)
+        for name, spec_parameter in spec_parameters.items():
+            if spec_parameter.required:
+                mandatory_params.add(name)
+                validation_params.add(name)
+            else:
+                validation_params.add(name)
+
+        extra_params = provided_params - (validation_params | mandatory_params)
+        if extra_params:
+            raise BadRequestError(msg=fmt_error('Invalid', extra_params))
+
+        if mandatory_params:
+            missing_params = mandatory_params - provided_params
+            if missing_params:
+                # Sorting is to produce a deterministic error message
+                raise BadRequestError(msg=fmt_error('Missing required', sorted(missing_params)))
+
+        provided_params &= validation_params
+        for param_name in provided_params:
+            param_value = all_provided_params[param_name]
+            spec_parameters[param_name].validate(dict(name=param_name, value=param_value))
+
+    def _extract_parameters(self, resource_path: str, http_method: str) -> List[JSON]:
+        """ Returns a list of OpenAPI `parameters` for a given resource_path from the api_spec"""
+        http_method = http_method.lower()
+        parameters = list()
+        parameters.extend(self.specs['paths'][resource_path].get('parameters', []))
+        parameters.extend(self.specs['paths'][resource_path].get(http_method, {}).get('parameters', []))
+        return parameters
+
+    def route(self,
```

Comment:
This method is mostly just duplicated from the super class. Try and see if you can find a way to reduce that with a call to `super().route()`. It may involve refactoring the parent class's `__init__`.

Reply:
within the previous refactor i was trying to do something like;

```
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/azul/chalice.py	(revision a1bd2ef5517f26f5dd7c10080d0b24094fe89c3d)
+++ src/azul/chalice.py	(date 1597947274672)
@@ -2,6 +2,7 @@
 import json
 import logging
 from typing import (
+    Callable,
     Dict,
     Iterable,
     List,
@@ -59,6 +60,7 @@
               enabled: bool = True,
               path_spec: Optional[JSON] = None,
               method_spec: Optional[JSON] = None,
+              wrapped_function: Optional[Callable] = False,
               **kwargs):
         """
         Decorates a view handler function in a Chalice application.
@@ -90,15 +92,23 @@
 
             def decorator(view_func):
                 self._register_spec(path, path_spec, method_spec, methods)
+
                 # Stash the URL path a view function is bound to as an attribute of
-                # the function itself.
-                view_func.path = path
-                return chalice_decorator(view_func)
+                # the function itself
+                @functools.wraps(view_func)
+                def f(*args, **kwargs):
+                    if wrapped_function:
+                        wrapped_function()
+                    return view_func(*args, **kwargs)
+
+                f.path = path
+                return chalice_decorator(f)
 
             return decorator
         else:
             return lambda view_func: view_func
 
+
     def test_route(self, *args, **kwargs):
         """
         A route that's only enabled during unit tests.
@@ -231,50 +241,10 @@
               method_spec: Optional[JSON] = None,
               validate: bool = True,
               **kwargs):
-        """
-        Decorates a view handler function in a Chalice application.
-
-        See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route.
-
-        :param path: See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route
-
-        :param path_spec: Corresponds to an OpenAPI Paths Object. See
-                          https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#pathsObject
-                          If multiple `@app.route` invocations refer to the same
-                          path (but with different HTTP methods), only specify
-                          this argument for one of them, otherwise an
-                          AssertionError will be raised.
-
-        :param method_spec: Corresponds to an OpenAPI Operation Object. See
-                            https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#operationObject
-                            This should be specified for every `@app.route`
-                            invocation.
-
-
-        :param enabled: If False, do not route any requests to the decorated
-                        view function. The application will behave as if the
-                        view function wasn't decorated.
-
-        :param validate: If False, do not validate requests submitted to the endpoint.
-        """
-        if enabled:
-            methods = kwargs.get('methods', ())
-            chalice_decorator = super().route(path, **kwargs)
-
-            def decorator(view_func):
-                self._register_spec(path, path_spec, method_spec, methods)
-
-                # Stash the URL path a view function is bound to as an attribute of
-                # the function itself
-                @functools.wraps(view_func)
-                def f(*args, **kwargs):
-                    if validate:
-                        self.validate_params()
-                    return view_func(*args, **kwargs)
-
-                f.path = path
-                return chalice_decorator(f)
-
-            return decorator
-        else:
-            return lambda view_func: view_func
+        kwargs_to_pass = dict(path=path,
+                              enabled=enabled,
+                              path_spec=path_spec,
+                              method_spec=method_spec)
+        if validate:
+            kwargs_to_pass['wrapped_function'] = self.validate_params
+        super().route(**kwargs_to_pass)
\ No newline at end of file

```

But kept running into:
```
Failure
Traceback (most recent call last):
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/suite.py", line 163, in _handleClassSetUp
    setUpClass()
  File "/Users/amar/dev/hca/azul/test/indexer/__init__.py", line 61, in setUpClass
    super().setUpClass()
  File "/Users/amar/dev/hca/azul/test/es_test_case.py", line 38, in setUpClass
    super().setUpClass()
  File "/Users/amar/dev/hca/azul/test/app_test_case.py", line 88, in setUpClass
    cls.app_module = load_app_module(cls.lambda_name(), unit_test=True)
  File "/Users/amar/dev/hca/azul/src/azul/modules.py", line 45, in load_app_module
    return load_module(path, f'lambdas.{lambda_name}.app', module_attributes)
  File "/Users/amar/dev/hca/azul/src/azul/modules.py", line 36, in load_module
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/amar/dev/hca/azul/lambdas/service/app.py", line 375, in <module>
    def swagger_ui():
TypeError: 'NoneType' object is not callable
```
ill take another shot at this.

Reply:
Here's a version: 
```
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/azul/chalice.py	(revision a1bd2ef5517f26f5dd7c10080d0b24094fe89c3d)
+++ src/azul/chalice.py	(date 1598312031206)
@@ -2,10 +2,12 @@
 import json
 import logging
 from typing import (
+    Callable,
     Dict,
     Iterable,
     List,
     Mapping,
+    MutableSequence,
     Optional,
 )
 
@@ -52,6 +54,7 @@
             self.specs['paths'] = {}
         else:
             self.specs: Optional[MutableJSON] = None
+        self.pre_call_hooks: MutableSequence[Callable[[], None]] = []
         super().__init__(app_name, debug=config.debug > 0, configure_logs=False)
 
     def route(self,
@@ -90,6 +93,8 @@
 
             def decorator(view_func):
                 self._register_spec(path, path_spec, method_spec, methods)
+                for hook in self.pre_call_hooks:
+                    hook()
                 # Stash the URL path a view function is bound to as an attribute of
                 # the function itself.
                 view_func.path = path
@@ -224,57 +229,10 @@
         parameters.extend(self.specs['paths'][resource_path].get(http_method, {}).get('parameters', []))
         return parameters
 
-    def route(self,
-              path: str,
-              enabled: bool = True,
-              path_spec: Optional[JSON] = None,
-              method_spec: Optional[JSON] = None,
-              validate: bool = True,
-              **kwargs):
+    def route(self, validate: bool = True, **kwargs):
         """
-        Decorates a view handler function in a Chalice application.
-
-        See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route.
-
-        :param path: See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route
-
-        :param path_spec: Corresponds to an OpenAPI Paths Object. See
-                          https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#pathsObject
-                          If multiple `@app.route` invocations refer to the same
-                          path (but with different HTTP methods), only specify
-                          this argument for one of them, otherwise an
-                          AssertionError will be raised.
-
-        :param method_spec: Corresponds to an OpenAPI Operation Object. See
-                            https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#operationObject
-                            This should be specified for every `@app.route`
-                            invocation.
-
-
-        :param enabled: If False, do not route any requests to the decorated
-                        view function. The application will behave as if the
-                        view function wasn't decorated.
-
         :param validate: If False, do not validate requests submitted to the endpoint.
         """
-        if enabled:
-            methods = kwargs.get('methods', ())
-            chalice_decorator = super().route(path, **kwargs)
-
-            def decorator(view_func):
-                self._register_spec(path, path_spec, method_spec, methods)
-
-                # Stash the URL path a view function is bound to as an attribute of
-                # the function itself
-                @functools.wraps(view_func)
-                def f(*args, **kwargs):
-                    if validate:
-                        self.validate_params()
-                    return view_func(*args, **kwargs)
-
-                f.path = path
-                return chalice_decorator(f)
-
-            return decorator
-        else:
-            return lambda view_func: view_func
+        if validate:
+            self.pre_call_hooks.append(self.validate_params)
+        super().route(**kwargs)
```

To take this idea to it's extreme, you could make `self._register_spec` be a `pre_call_hook` too. Then the signature for hooks would have to be changed to accept everything that's passed in to the route method. While this version would be more powerful and extendable, it might also be overkill.

Reply:
This seems to be in-lining callables at the same time the registration of the routes occur. We need to be performing the validation at a per-request basis, so it would have to be when the method that the route calls is invoked.

I'll take a look at this at a later time and focus on merging more code into the PathSpec class.